File: C/Program/bfs.c
#include <stdlib.h>
#include <stdio.h>
#include "generator.h"
#include "bfs.h"

void addToQueue (t_queue * queue, int vertex)
{
    queue->cells[queue->end] = vertex;
    queue->end++;
}

int popFromQueue (t_queue * queue)
{
    if (queue->start < queue->end)
    {
        int value = queue->cells[queue->start];
        queue->start++;
        return value;
    }
    else
    {
        return -2;
    }
}


int BFS (t_pair ** graph, int n, int startingVertex)
{
    int *vertexState = malloc (n * (sizeof (int)));	//creating array that will keep information whether vertex has been visited
    for (int i = 0; i < n; i++)
    {
        //initially all vertexes are unvisitied (vertexState=0)
        vertexState[i] = 0;
    }
    t_queue * queue = malloc (sizeof (t_queue));
    queue->cells = malloc ((n + 3) * (sizeof (int)));
    queue->start = 0;
    queue->end = 0;

    addToQueue (queue, startingVertex);
    int currentVertex;
    while ((currentVertex = popFromQueue (queue)) != -2)
    {
        //while queue isn't empty
        for (int j = 0; j < 4; j++)
        {
            if (vertexState[graph[currentVertex][j].vertex] == 0
                && graph[currentVertex][j].vertex != -1)
            {
                addToQueue (queue, graph[currentVertex][j].vertex);
                vertexState[graph[currentVertex][j].vertex] = 1;	//1 - added to queue
            }
        }
        vertexState[currentVertex] = 2;	//visited
    }

    int haveAllVertexesBeenVisited = 1;	//1 - they have
    for (int i = 0; i < n; i++)
    {
        if (vertexState[i] != 2)
        {
            haveAllVertexesBeenVisited = 0;	//0 - they haven't
            break;
        }
    }
    return haveAllVertexesBeenVisited;
}


int isConst (t_pair ** graph, int rows, int columns)
{
    int wynik = 1;

    for (int i = 0; i < (rows * columns); i++)
    {
        wynik = BFS (graph, (rows * columns), i);
        if (wynik == 0) break;
    }

    if (wynik == 1)
    {
        printf ("\n\n Wynik dzialania BFSa: spojny\n");
        return 0;
    }

    else
    {
        printf ("\n\n Wynik dzialania BFSa: niespojny\n");
        return 1;
    }
}

File: Java/JIMPGraph/src/main/java/graph/jimpgraph/Bfs.java
package graph.jimpgraph;

import java.util.LinkedList;
import java.util.Queue;

public class Bfs {

    static LinkedList<Integer> queue;
    static int[] vertexState;

    public static int BFS(Graph graph, int startingVertex){
        int n = graph.columns*graph.rows;
        queue = new LinkedList<Integer>();
        vertexState = new int[n];
        for(int i=0;i<n;i++){
            vertexState[i] = 0;
        }
        queue.add(startingVertex);
        int currentVertex;
        while(!queue.isEmpty()){
            currentVertex = queue.removeFirst();
            for (int j = 0; j < graph.directions; j++)
            {
                if (graph.vertex[currentVertex][j] != -1) {
                    if (vertexState[graph.vertex[currentVertex][j]] == 0) {
                        queue.add(graph.vertex[currentVertex][j]);
                        vertexState[graph.vertex[currentVertex][j]] = 1; // 1 - added to queue
                    }
                }
            }
            vertexState[currentVertex] = 2; // visited


        }
        int haveAllVertexesBeenVisited = 1;
        for (int i = 0; i < n; i++)
        {
            if (vertexState[i] != 2)
            {
                haveAllVertexesBeenVisited = 0;
                break;
            }
        }
        return haveAllVertexesBeenVisited;

    }


    public static String isConstCommunicate (Graph graph){
        if(BFS(graph,0) == 1) return "Graf jest spójny!";
        else return "Graf jest niespójny!";
    }
}
